mettle/src/stdapi/fs/file.c

```diff
@@ -160,6 +167,274 @@ struct tlv_packet *fs_ls(struct tlv_handler_ctx *ctx)
 	return NULL;
 }
 
+static void
+search_add_result(struct tlv_handler_ctx *ctx, struct tlv_packet **p, char *sub_root, char *f_name, off_t f_size)
+{
+	struct tlv_packet *res = NULL;
+
+	if(!*p)
+	{
+		*p = tlv_packet_response_result(ctx, TLV_RESULT_SUCCESS);
+	}
+
+	res = tlv_packet_new(TLV_TYPE_SEARCH_RESULTS, 0);
+	res = tlv_packet_add_str(res, TLV_TYPE_FILE_PATH, sub_root);
+	res = tlv_packet_add_str(res, TLV_TYPE_FILE_NAME, f_name);
+	res = tlv_packet_add_u32(res, TLV_TYPE_FILE_SIZE, f_size);
+	*p = tlv_packet_add_child(*p, res);
+}
+
+#ifdef HAVE_GLOB
+static int
+search_glob(struct tlv_handler_ctx *ctx, struct tlv_packet **p, char *sub_root, char *f_name)
+{
+	glob_t glob_results;
+	struct stat s_buf;
+	char glob_path[PATH_MAX + 1];
+
+	if(snprintf(glob_path, PATH_MAX + 1, "%s/%s", sub_root, f_name) < 0)
+	{
+		return -1;
+	}
+
+#ifndef GLOB_TILDE
+#define GLOB_TILDE 0
+#endif
+
+#ifndef GLOB_PERIOD
+#define GLOB_PERIOD 0
+#endif
+
+	memset(&glob_results, 0, sizeof(glob_t));
+	memset(&s_buf, 0, sizeof(struct stat));
+	if(glob(glob_path, GLOB_TILDE | GLOB_PERIOD, NULL, &glob_results) != 0)
+	{
+		globfree(&glob_results);
+		return -1;
+	}
+
+	for(size_t i = 0; i < glob_results.gl_pathc; i++)
+	{
+#ifdef _WIN32
+		if(stat(glob_results.gl_pathv[i], &s_buf) == 0)
+#else
+		if(lstat(glob_results.gl_pathv[i], &s_buf) == 0)
+#endif
+		{
+			search_add_result(ctx, p, sub_root, basename(glob_results.gl_pathv[i]), s_buf.st_size);
+		}
+	}
+
+	globfree(&glob_results);
+	return 0;
+}
+#endif
+
+static void
+fs_search_cb(eio_req *req)
+{
+	bool recurse;
+	bool perform_glob = false;
+	int rc = TLV_RESULT_SUCCESS;
+	struct tlv_packet *p = NULL;
+	struct tlv_handler_ctx *ctx = req->data;
+	char *path = tlv_packet_get_str(ctx->req, TLV_TYPE_SEARCH_GLOB);
+	char *search_root = tlv_packet_get_str(ctx->req, TLV_TYPE_SEARCH_ROOT);
+
+	DIR *dir_str;
+	struct dirent *f_entry;
+	struct search_entry *curr_entry;
+
+	tlv_packet_get_bool(ctx->req, TLV_TYPE_SEARCH_RECURSE, &recurse);
+
+	if(search_root == NULL || (strcmp(search_root, "") == 0))
+	{
+		search_root = ".";
```

Comment:
It's worth noting that on Windows if the `-d` option is omitted, the search root appears to search all drives:
https://github.com/rapid7/metasploit-payloads/blob/master/c/meterpreter/source/extensions/stdapi/server/fs/search.c#L850
I guess this behaviour kind of makes sense on Windows (because you can't use / to search everywhere), but I prefer the behaviour here (searching the current directory). :shrug:

Reply:
On second thoughts I think I'm going to change this to '/' and land it, for consistency (and to avoid updating `meterpreter search -h`). But I'm easily persuaded to change it back.

Reply:
I tried modeling this after python meterpreter's search functionality which uses the current directory by default. I prefer using the current directory as well, but it seems far more trivial to just use `/` for this and python meterpreter versus changing the Windows meterpreter functionality, so I'll just PR a change for the python side now. Thanks for handling this!